src/test/java/org/jenkinsci/plugins/displayurlapi/DisplayURLProviderTest.java

```diff
@@ -48,6 +56,163 @@ public void urls() throws Exception {
+                     DisplayURLProvider.get().getArtifactsURL(run));
+        assertEquals(root + "job/my%20folder/job/my%20job/changes/" + TestSysPropDisplayURLProvider.EXTRA_CONTENT_IN_URL,
+                     DisplayURLProvider.get().getChangesURL(run));
+        assertEquals(root + "job/my%20folder/job/my%20job/1/testReport/" + TestSysPropDisplayURLProvider.EXTRA_CONTENT_IN_URL,
+                     DisplayURLProvider.get().getTestsURL(run));
+
+        EnvVars environment = run.getEnvironment();
+        assertEquals(DisplayURLProvider.get().getRunURL(run), environment.get("RUN_DISPLAY_URL"));
+        assertEquals(DisplayURLProvider.get().getArtifactsURL(run), environment.get("RUN_ARTIFACTS_DISPLAY_URL"));
+        assertEquals(DisplayURLProvider.get().getChangesURL(run), environment.get("RUN_CHANGES_DISPLAY_URL"));
+        assertEquals(DisplayURLProvider.get().getTestsURL(run), environment.get("RUN_TESTS_DISPLAY_URL"));
+        assertEquals(DisplayURLProvider.get().getJobURL(project), environment.get("JOB_DISPLAY_URL"));
+        System.setProperty("jenkins.displayurl.provider", "");
+    }
+
+    @TestExtension
+    public static class TestUserDisplayURLProvider extends DisplayURLProvider
+    {
+
+        public static final String EXTRA_CONTENT_IN_URL = "userdefinedprovider";
+
+        private DisplayURLProvider getRootProvider()
+        {
+            return ExtensionList.lookup( DisplayURLProvider.class ).getDynamic(
+                ClassicDisplayURLProvider.class.getName() );
+        }
+
+        @Override
+        public String getRunURL( Run<?, ?> run )
+        {
+            return getRootProvider().getRunURL( run ) + EXTRA_CONTENT_IN_URL;
+        }
+
+        @Override
+        public String getArtifactsURL( Run<?, ?> run )
+        {
+            return getRootProvider().getArtifactsURL( run ) + "/" + EXTRA_CONTENT_IN_URL;
+        }
+
+        @Override
+        public String getChangesURL( Run<?, ?> run )
+        {
+            return getRootProvider().getChangesURL( run ) + "/" + EXTRA_CONTENT_IN_URL;
+        }
+
+        @Override
+        public String getTestsURL( Run<?, ?> run )
+        {
+            return getRootProvider().getTestsURL( run ) + "/" + EXTRA_CONTENT_IN_URL;
+        }
+
+        @Override
+        public String getJobURL( Job<?, ?> project )
+        {
+            return getRootProvider().getJobURL( project ) + EXTRA_CONTENT_IN_URL;
+        }
+    }
+
+    @Test
+    public void urlsWithUserDefinedProvider() throws Exception {
+        User foo = User.getById("foo", true);
+        PreferredProviderUserProperty userProperty =
+            new PreferredProviderUserProperty(TestUserDisplayURLProvider.class.getName());
+        foo.addProperty(userProperty);
+        Authentication previous = SecurityContextHolder.getContext().getAuthentication();
+        SecurityContextHolder.getContext().setAuthentication(new UsernamePasswordAuthenticationToken(foo, null));
```

Comment:
there is maybe something easier to change current user but I was not sure....

Reply:
I think `try (ACLContext c = ACL.as(foo)) { ... code to run with foo as the current user ... }` is the recommended approach.